### DISCUSSION TITLE
Using Crypto API

### DISCUSSION THREAD

**Discussion by idleberg** [vitest-dev/vitest#893] [Q&A] [answered]
Does anybody here have experience on how to run tests that involve the Crypto API? In a React project, I have a function creating a hash from a string.

```js
const encoder = new TextEncoder();
const data = encoder.encode(message);
const hash = await crypto.subtle.digest("SHA-256", data);
```

However, when I try and test this function using Vitest, I'm running into a type error:

> Cannot read properties of undefined (reading 'digest')

I already added the following to my tests:

```js
import crypto from "node:crypto";
globalThis.crypto = crypto;
```

Without it, it already fails at `crypto`.

Can anyone please help?

**Comment by sheremet-va** [comment]
Try 
```
globalThis.crypto = crypto
globalThis.window.crypto = crypto
```

**Reply by idleberg** [reply]
> globalThis.crypto = crypto
> globalThis.window.crypto = crypto

Thanks, but unfortunately it doesn't change anything

**Comment by idleberg** [comment, marked as answer]
Turns out I need Node 17+ to get the browser-compatible `crypto` API

**Reply by KevinGhadyani-Okta** [reply]
I've upgraded to the latest `vitest` on two completely different projects now, and both have the `crypto.hash` error on Node `v20.11.0`, so I don't think it's just having a newer version of Node.
